flow-typed/npm/typescript_v3.3.x.js

```diff
@@ -8444,7 +8444,7 @@ declare module "typescript" {
     +getProjectReferences?: () => $ReadOnlyArray<ProjectReference> | void,
     +getLocalizedDiagnosticMessages?: () => any,
     +getCancellationToken?: () => HostCancellationToken,
-    getCurrentDirectory(): string,
+    getCurrentDirectory: () => string,
```

Comment:
The "methods" for this interface are a mix of methods and properties whose values are arrow functions. Methods cannot be unbound, but properties can, making them incompatible — unfortunately this means that our subclass can't use entirely methods nor properties.

I think the reasoning for this is the interfaces these inherit from have optional properties, something that can't be done with methods, and since they're incompatible, they must stay consistent. That's the case with `getCurrentDirectory`, which is a property in the parent interface but was a method here.

@mischnic do you have any ideas how we could make these more consistent? Or if this change should stay or be submitted to flow-typed?

Reply:
This definitely needs to be upstreamed because right now the libdef is just broken 🤷 

There are code patterns that cannot be modeled with Flow (either unsound with arrow properties because the implementation of `run` shouldn't be unbound, or you cannot model the optional method):

```js
interface Super {
  /* "optional" */ run(): string;
}

// should pass
class Super2 implements Super {
}

const s1 = new Super1();

// should fail
s1.run
```

You can only make `run` an arrow function if the implementation (= tsc) doesn't rely on `this` at runtime.